### DISCUSSION TITLE
Using SQLAlchemy in both sync and async code

### DISCUSSION THREAD

**Discussion by zikphil** [sqlalchemy/sqlalchemy#5923] [Usage Questions] [answered]
Hello community,

I would like to pick your brain with my current dilemma. My current architecture includes an async api using Tornado, a shared library that includes all my SQLAlchemy models and custom query functions and a Celery 5.0 sync worker.

I am trying to find a way that I could re-use my shared library no matter the code base, so in other words finding a way to use my SQLAlchemy shortcut query functions in both sync and async code. Here are the options I am considering:

1. Moving my query functions away from the Model class and build dedicated DAL-like classes to hold my custom queries with a custom __getattr__ that would convert my async query into sync upon execution when I use them in my Celery worker.
2. Only hold the query itself in the functions, not involving the session and instead manually execute the query with the proper SyncSession or AsyncSession depending on who is calling them.
3. Have duplicated code that have both a async and async version of the functions.

Anybody have found themselves in a similar situation and could share how they did it?

Thanks,

Phil

**Comment by zzzeek** [comment, marked as answer]
Python asyncio APIs are interchangeable with sync APis in both directions.    so implement your API in one or the other, then use one of the below methods to translate:


Async API to Sync API


```python


async def api_function_one():
    print("hi from af one()")
    await api_function_two()
    return "result"

async def api_function_two():
    print("hi from af two()")


def sync_api_function_one():
    coro = api_function_one()
    try:
        coro.send(None)
    except StopIteration as err:
        return err.value


print(sync_api_function_one())
```

Sync API to Async API

```python
import asyncio

from sqlalchemy.util import concurrency


def api_function_one():
    print("hi from af one()")
    api_function_two()
    return "result"


def api_function_two():
    print("hi from af two()")


async def async_api_function_one():
    return await concurrency.greenlet_spawn(api_function_one)


print(asyncio.run(async_api_function_one()))

```



the two approaches seem pretty easy, what's the catch?   It's that when the API functions do anything involving IO, that the IO method in use is aware of whether or not we are running in async or non-async.   In SQLAlchemy, this is achieved by use of the SQLAlchemy dialects, which are hardcoded to be either sync or async.     If your API code is dealing just with SQLAlchemy sync or async APIs, you will be good.    the application that runs at the top level as asyncio should use an async dialect like asyncpg, and the application that runs at the top level as sync should use a sync driver like psycopg2.

The main idea is that the API examples above represent **the middle part between someone's application and an IO driver of some kind**.    Mediating between Python's "async" keywords vs. not is a matter of one of the above two techniques.

is this hard to understand?  absolutely.  But IMO that's all things asyncio :)  .

**Reply by zikphil** [reply]
Following up on that, the first use case is what applies to me. Everything seems to work great except for the signals! Remember our previous discussion here: https://github.com/sqlalchemy/sqlalchemy/issues/5905

When I have the following signal:
```
@event.listens_for(User, 'after_update')
@sync_as_async
async def user_post_save(mapper, connection, target):
    print('Hello Cruel World')
```

It works great when I have an AsyncSession commit User changes but when I have a SyncSession call it, I get the following error and it does not even print the text:

```
...
  File "/Users/phil/.pyenv/versions/nvp/lib/python3.8/site-packages/sqlalchemy/orm/unitofwork.py", line 621, in execute
    util.preloaded.orm_persistence.save_obj(
  File "/Users/phil/.pyenv/versions/nvp/lib/python3.8/site-packages/sqlalchemy/orm/persistence.py", line 250, in save_obj
    _finalize_insert_update_commands(
  File "/Users/phil/.pyenv/versions/nvp/lib/python3.8/site-packages/sqlalchemy/orm/persistence.py", line 1540, in _finalize_insert_update_commands
    mapper.dispatch.after_update(mapper, connection, state)
  File "/Users/phil/.pyenv/versions/nvp/lib/python3.8/site-packages/sqlalchemy/event/attr.py", line 320, in __call__
    fn(*args, **kw)
  File "/Users/phil/.pyenv/versions/nvp/lib/python3.8/site-packages/sqlalchemy/orm/events.py", line 743, in wrap
    fn(*arg, **kw)
  File "/Volumes/Projects/nrw/backend/nvp/utils/sqla.py", line 7, in go
    return await_only(fn(*arg, **kw))
  File "/Users/phil/.pyenv/versions/nvp/lib/python3.8/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 47, in await_only
    raise exc.MissingGreenlet(
sqlalchemy.exc.MissingGreenlet: greenlet_spawn has not been called; can't call await_() here. Was IO attempted in an unexpected place? (Background on this error at: http://sqlalche.me/e/14/xd2s)
```

The moment I change the signal to:
```
@event.listens_for(User, 'after_update')
def user_post_save(mapper, connection, target):
    print('Hello Cruel World')
```

it works fine.

Does that mean that it would be better if Signals are always SYNC or is that a bug somewhere? I assume that when SQLAlchemy is calling the signal it is trying to get the connection object and in this case it excepts an AsyncObject? Because there are no SQLAlchemy code in this signal and I still get the error message.

**Reply by zzzeek** [reply]
it would be better if you kept your event handlers as sync, because they are always run inside the "sync" part of the library, and below I illustrate why it's really quite wasteful to go back into "async" on the inside, but it is doable still.

as far as why your handler is failing, it's because the sync_as_async demo I gave you assumes you are invoking everything from the "async" API, which means 1. there's an asyncio event loop running and 2. the SQLAlchemy asyncio extension is in use which means all calls into SQLAlchemy's "sync" API occur within greenlet_spawn.

Overall, this is the thing that's going on:  there is Python "async" world, where they want you to have async/await for everything, and "sync" world, where they dont.      You can transition between these two worlds using the two patterns I have above.   But if you only use part of one of the transitions, it wont work:

1. async/await -> can call async/await IO methods
2. Normal sync code -> can call sync IO methods
3. async /await ->   [SQLAlchemy's asyncio APIs: can call greenlet_spawn] -> [SQLAlchemy's asyncio DBAPI drivers: can call await_only_ ] ->  can call async/await IO methods

the sync_as_async thing is doing this:

4. async /await ->  [SQLAlchemy's asyncio APIs: can call greenlet_spawn]  -> [the decorator I gave you: can call await_only_] -> [your async event function: can call async/await ] -> [back into SQLALchemy asyncio: -> can call greenlet_spawn] ->  [SQLAlchemy's asyncio DBAPI drivers: can call await_only_ ] ->  can call async/await IO methods

So above, that's a lot of contextual switching and it will add measurable overhead.  But you are inside a single event loop the whole time, so there's that.

When you say you want to run SyncSession, but then have event handlers that use async defs, the approach that would make this possible depends on what kind of concurrency model is in use in the first place.     if the use of SyncSession implies that you are working in threaded code against a blocking database driver, then you would alter the decorator as follows:

```python
import greenlet
from sqlalchemy.util.concurrency import await_only
from sqlalchemy.util._concurrency_py3k import _AsyncIoGreenlet

def running_in_greenlet():
    return isinstance(greenlet.getcurrent(), _AsyncIoGreenlet)

def sync_as_async(fn):
    def go(*arg, **kw):
        if running_in_greenlet():
            return await_only(fn(*arg, **kw))
        else:
            # no greenlet, assume no event loop and blocking IO backend
            coro = fn(*arg, **kw):
            try:
                coro.send(None)
            except StopIteration as err:
                return err.value
    return go

```

if OTOH a method is called on SyncSession and there's an asyncio database driver in use, I would say that there needs to be an event loop already running for the approach to make any sense which means any calls to SyncSession should be within the greenlet.   There's another function called await_fallback() that will spin up an event loop for every call which would allow this scenario to "work" but it renders the use of asyncio as pointless; it's something we were using for testing only.

**Reply by nburns** [reply]
I found that I actually need to call `asyncio.get_event_loop().run_until_complete` instead of `asyncio.run` because the extra event loop from `asyncio.run` https://github.com/python/cpython/blob/3.10/Lib/asyncio/runners.py#L39 prevents the session from being returned to the pool
